atomic_reactor/util.py

```diff
@@ -1557,3 +1558,64 @@ def has_operator_manifest(workflow):
     except KeyError:
         operator_label = 'false'
     return operator_label.lower() == 'true'
+
+
+def annotation(*keys):
+    """
+    Annotate a `BuildPlugin` subclass. The `run()` method of this plugin will
+    store its results in the plugin's workflow as annotations.
+
+    The `run()` method of the plugin being annotated has to return a dict
+    containing exactly the specified keys.
+
+    The `store_metadata_in_osv3` plugin will automatically collect these
+    annotations and upload them to OpenShift.
+
+    Example:
+    >>> from atomic_reactor.plugin import BuildPlugin
+
+    >>> @annotation('foo', 'bar')
+    >>> class MyBuildPlugin(BuildPlugin):
+    >>>     key = 'my_build_plugin'
+    >>>
+    >>>     def run(self):
+    >>>         return {'foo': 1, 'bar': 2}
+
+    >>> p = MyBuildPlugin(...)
+    >>> p.run()
+    >>> p.workflow.annotations  # {'foo': 1, 'bar': 2}
+
+    :param keys: Keys to annotate the plugin with
+    :return: Decorator that will turn the plugin into an annotated one
+    """
+    from atomic_reactor.plugin import BuildPlugin
+
+    def update_annotations(annotations, result):
+        if set(keys) != set(result):
+            msg = ('Expected the following annotations: {}, got: {}'
+                   .format(sorted(keys), sorted(result)))
+            raise TypeError(msg)
```

Comment:
Would we want to allow the result to have keys other than those expected by the annotation? If not, then specifying the keys in the decorator is a bit redundant but it might serve as an extra bit of "self-documentation".

Reply:
I can *definitely* see it being useful to return other keys than those we want an annotation for. Some plugins need to communicate to others, but we don't necessarily want to clutter up the Build annotations with that chatter (annotation size is limited).

Reply:
Right, I suspected that

Reply:
Also updated the example in docstring to reflect this